carapace-server/src/main/java/org/carapaceproxy/server/certificates/Route53Client.java

```diff
@@ -0,0 +1,166 @@
+        performActionOnDNSChallengeForDomain(domain, digest, DELETE, onRequestComplete, onRequestFail);
+    }
+
+    public void isDNSChallengeForDomainAvailable(String domain, String digest, Consumer<Boolean> onRequestComplete, Consumer<Throwable> onRequestFail) {
+        Consumer<Route53Response> check = res -> {
+            TestDnsAnswerResponse r = (TestDnsAnswerResponse) res;
+            onRequestComplete.accept(r.recordData().size() == 1 && r.recordData().get(0).equals("\"" + digest + "\""));
+        };
+        performActionOnDNSChallengeForDomain(domain, null, CHECK, check, onRequestFail);
+    }
+
+    private void performActionOnDNSChallengeForDomain(String domain, String digest, DNSChallengeAction action, Consumer<Route53Response> onRequestComplete,
+                                                      Consumer<Throwable> onRequestFail) {
+        String challengeName = DNS_CHALLENGE_PREFIX + domain;
+
+        // find out the hostedzone where to update the acme dns-challenge txt record
+        CompletableFuture<ListHostedZonesByNameResponse> futureFindHZ = client.listHostedZonesByName(
+                ListHostedZonesByNameRequest.builder().dnsName(domain).build()
+        );
+
+        futureFindHZ.whenComplete((resFindHZ, excFindHZ) -> {
+            if (resFindHZ != null && resFindHZ.sdkHttpResponse().isSuccessful() && !resFindHZ.hostedZones().isEmpty()) {
+                String idhostedzone = resFindHZ.hostedZones().get(0).id().replace(HOSTEDZONE_ID_PREFIX, "");
+                CompletableFuture<? extends Route53Response> future;
+                switch (action) {
+                    case UPSERT:
+                        // ACME DNS-challenge TXT record upsert
+                        future = client.changeResourceRecordSets(
+                                acmeDNSChallengeUpdateRequest(idhostedzone, challengeName, digest, ChangeAction.UPSERT)
+                        );
+                        break;
+                    case DELETE:
+                        // ACME DNS-challenge TXT record delete
+                        future = client.changeResourceRecordSets(
+                                acmeDNSChallengeUpdateRequest(idhostedzone, challengeName, digest, ChangeAction.DELETE)
+                        );
+                        break;
+                    case CHECK:
+                        // ACME DNS-challenge TXT record availability check
+                        future = client.testDNSAnswer(TestDnsAnswerRequest.builder()
+                                .hostedZoneId(idhostedzone)
+                                .recordName(challengeName)
+                                .recordType(RRType.TXT)
+                                .build()
+                        );
+                        break;
+                    default:
+                        throw new IllegalArgumentException("Unexpected value for action: " + action);
+                }
+                future.whenComplete((res, exc) -> {
```

Comment:
you have to wait for this future to complete
the one returned by future.whenComplete